File: Greedy/MaximumTrains/MaximumTrains.js
function method1(arr) {
  let nop = 0;
  arr.sort((a1, a2) => {
    nop = Math.max(nop, Math.max(a1[2], a2[2]));
    return a1[1] - a2[1];
  });
  let platform = new Array(nop + 1);
  platform.fill(-1);
  let count = 0;
  for (let ar of arr) {
    if (platform[ar[2]] == -1) {
      count++;
      platform[ar[2]] = ar[1];
    } else {
      if (platform[ar[2]] < ar[0]) {
        count++;
        platform[ar[2]] = ar[1];
      }
    }
  }
  return count;
}

let arr = [
  [1000, 1030, 1],
  [1010, 1030, 1],
  [1000, 1020, 2],
  [1030, 1230, 2],
  [1200, 1230, 3],
  [0900, 1005, 1],
];

console.log(method1(arr));

File: Greedy/MaximumTrains/MaximumTrains.go
package main

import (
	"fmt"
	"sort"
)

func method1(arr [][]int) int {
	sort.Slice(arr, func(i, j int) bool {
		return arr[i][1] < arr[j][1]
	})
	n := len(arr)
	max := 0
	for i := 0; i < n; i++ {
		if arr[i][2] > max {
			max = arr[i][2]
		}
	}
	count := 0
	list := make([]int, max+1)
	for i := 0; i < n; i++ {
		plat := arr[i][2]
		start := arr[i][0]
		end := arr[i][1]
		if list[plat] == 0 {
			count++
			list[plat] = end
		} else {
			if list[plat] < start {
				count++
				list[plat] = end
			}
		}
	}
	return count
}

func main() {
	arr := [][]int{
		{1000, 1030, 1},
		{1010, 1030, 1},
		{1000, 1020, 2},
		{1030, 1230, 2},
		{1200, 1230, 3},
		{900, 1005, 1},
	}
	fmt.Println(method1(arr))
	//fmt.Println(arr)
}
